### PR TITLE
Rework poi

### LINKED ISSUE #1905: Rework proof of index
### Description
Currently we offer two levels of cryptographic proof for index, 
1. `hash` in the poi table, which is `HashFn(height,chainBlockHash,parentHash,operationHash,projectId)`
    1. poi table is not continuous (only when there is operation in the block)
    2. parentHash is the hash of last block with operation
    3. if a block is skipped, we use a fixed hash for that block
2. We also build a mmr and calculate the root hash, because this proof is continuous we use if for the final proof

Theoretical, hash is enough to tell if two indexers are generating same data (because parenthash is included in the digest)
the only problem for it is not continuous.
So we can actually simplify poi by removing mmr and only use the hash, as long as we can solve the only blocking issue.

### Details
1. Make data in _poi table continuous
    1. We can keep the current place we insert poi record with block of operations, but we should leave `hash` empty
    2. Change where we used to insert mmr table to insert poi table, fill in empty blocks in the table. This step we can generate `hash` for records inserted in step 1
    3. For skipped blocks, we use empty hash for chainblockHash and operationHash, but the parentHash is no more the last block with operation, we use its parent block literally.
2. further tidy up _poi table, deprecating parentHash column since it is redundant data and now can be easily found by id.
3. change api, add `/poi/` and deprecate `/mmrs/`, change response body.

**Local Environment:** [You can get this information from executing `subql version`.]
**Query Version:** [What is the version of the query service?]
**Indexer Version:** [What is the version of the indexer service?]
**Network Details:**
  * [Network]
  * [Block height, a block height where the issue is happening]
  * [Dictionary endpoint, if used]

### Steps to Reproduce

1. [First Step]
5. [Second Step]
6. [and so on...]

**Example project:** [A link to a minimal example that can reproduce the issue]

**Expected behavior:** [What you expected to happen]

**Actual behavior:** [What actually happened]

### Any other information

Is there any other information you would like to add?

### PR DESCRIPTION
**Pull request by jiqiang90**
# Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #1905 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] I have tested locally
- [x] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 1]
```diff
@@ -1,55 +1,261 @@
...
+    if (!this._latestSyncedPoi) {
+      throw new Error('_latestSyncedPoi in Poi service is not inited');
+    }
+    if (this._latestSyncedPoi.parentHash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, parentHash is missing`);
+    }
+    if (this._latestSyncedPoi.hash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, hash is missing`);
+    }
+    return <SyncedProofOfIndex>this._latestSyncedPoi;
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:60]: You could use a typeguard here which would avoid the cast:

```ts
function isSyncedProofOfIndex(poi?: ProofOfIndex | SyncedProofOfIndex): poi is SyncedProofOfIndex {
  return !!poi && (poi as SyncedProofOfIndex).parentHash && (poi as SyncedProofOfIndex).hash;
}
```

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 2]
```diff
@@ -1,55 +1,261 @@
...
+  get projectId(): string {
+    if (!this._projectId) {
+      throw new Error(`No project Id inited`);
+    }
+    return this._projectId;
+  }
+
+  get latestSyncedPoi(): SyncedProofOfIndex {
+    if (!this._latestSyncedPoi) {
+      throw new Error('_latestSyncedPoi in Poi service is not inited');
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:52]:
```suggestion
      throw new Error('_latestSyncedPoi in Poi service is not initalized');
```

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 3]
```diff
@@ -1,55 +1,261 @@
...
+    }
+    if (this._latestSyncedPoi.parentHash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, parentHash is missing`);
+    }
+    if (this._latestSyncedPoi.hash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, hash is missing`);
+    }
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:59]: What are the chances of these happening? If a user faced this error can we give them some instructions how to resolve the issue?

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 4]
```diff
@@ -1,55 +1,261 @@
...
+  async init(schema: string, projectId: string): Promise<void> {
+    this._poiRepo = this.storeCache.poi ?? undefined;
+    this._projectId = projectId;
+    const latestSyncedPoiHeight = await this.storeCache.metadata.find('latestSyncedPoiHeight');
+    if (latestSyncedPoiHeight !== undefined) {
+      const recordedPoi = await this.poiRepo.getPoiById(latestSyncedPoiHeight);
+      if (recordedPoi) {
+        await this.setLatestSyncedPoi(recordedPoi);
+      } else {
+        // TODO, could use sql to find latestSyncedPoiHeight in db
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:72]: Still a TODO?

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 5]
```diff
@@ -1,55 +1,261 @@
...
+            throw e;
+          }
+        }
+        logger.info(`Successful migrate Poi`);
+      }
+      // Remove and Change column from sequelize not work, it only applies to public schema
+      // https://github.com/sequelize/sequelize/issues/13365
+      // await this.poiRepo?.model.sequelize?.getQueryInterface().changeColumn(tableName,'mmrRoot',{})
+    } catch (e) {
+      throw new Error(`Failed to migrate poi table}`);
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:150]:
```suggestion
      throw new Error(`Failed to migrate poi table. {e}`);
```

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 6]
```diff
@@ -1,55 +1,261 @@
...
+          } catch (e) {
+            logger.error(`Migration poi failed with query: ${`query`}`);
+            throw e;
+          }
+        }
+        logger.info(`Successful migrate Poi`);
+      }
+      // Remove and Change column from sequelize not work, it only applies to public schema
+      // https://github.com/sequelize/sequelize/issues/13365
+      // await this.poiRepo?.model.sequelize?.getQueryInterface().changeColumn(tableName,'mmrRoot',{})
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:148]: Can we run raw sql?

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 7]
```diff
@@ -1,55 +1,261 @@
...
+          );
+        }
+      }
+
+      if (queries.length) {
+        for (const query of queries) {
+          try {
+            await this.poiRepo?.model.sequelize?.query(query, {type: QueryTypes.SELECT});
+          } catch (e) {
+            logger.error(`Migration poi failed with query: ${`query`}`);
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:140]:
```suggestion
            logger.error(`Migration poi failed with query: ${query}`);
```

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 8]
```diff
@@ -1,55 +1,261 @@
...
+      } else {
+        // TODO, could use sql to find latestSyncedPoiHeight in db
+        throw new Error(`Can not find latestSyncedPoiHeight ${latestSyncedPoiHeight}`);
+      }
+    }
+    await this.migratePoi(schema);
+  }
+
+  // This could use for a subcommand later
+  async migratePoi(schema: string): Promise<void> {
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:80]: Does this need to clean up the MMR table? Or file based MMR?

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 9]
```diff
@@ -1,55 +1,261 @@
...
+      }
+      this.poiRepo?.bulkUpsert(appendedBlocks);
+    }
+  }
+
+  private addDefaultPoiBlocks(endHeight: number, appendedBlocks: ProofOfIndex[]): void {
+    for (let i = this.latestSyncedPoi.id + 1; i <= endHeight; i++) {
+      const syncedPoiBlock = PoiBlock.create(i, null, null, this.projectId, this.latestSyncedPoi.hash);
+      appendedBlocks.push(syncedPoiBlock);
+      void this.setLatestSyncedPoi(syncedPoiBlock);
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:258]: Why is this void?

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 8 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:80]: It cleans up mmr table if exist when migration in the first time. 

For fileBased mmr, because we have removed relate flags for mmr, we could ask user remove it manually.

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 6 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:148]: Current implementation is using raw sql, I left this comment here due to this bug here, we could not alter poi table columns with sequelize method `getQueryInterface().changeColumn`

**Comment by stwiname** [comment]
It would be good for Poi Service to have coverage over 80%
Screenshot 2023-08-24 at 3 01 59 PM

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 10]
```diff
@@ -1,55 +1,261 @@
...
-    if (!this.latestPoiBlockHash || !isMainThread) {
-      const poiBlockHash = await this.fetchPoiBlockHashFromDb();
-      if (poiBlockHash === null || poiBlockHash === undefined) {
-        this.latestPoiBlockHash = DEFAULT_PARENT_HASH;
+  async syncPoi(exitHeight?: number, logging?: boolean): Promise<void> {
+    if (this.isSyncing) return;
+    this.isSyncing = true;
+    while (!this.isShutdown) {
+      if (!this._latestSyncedPoi) {
+        await delay(10);
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:176]: If _latestSyncedPoi is undefined, which means we haven't sync any Poi yet, no genesis poi been found.  It delay for genesis poi to be created, delay will stop after `ensureGenesisPoi`, https://github.com/subquery/subql/blob/bfe6de09f6743db219f5e46f2022bf41711feb3a/packages/node-core/src/indexer/poi/poi.service.ts#L160

File: packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts [hunk 11]
```diff
@@ -197,27 +190,21 @@ export abstract class BaseBlockDispatcher<Q extends IQueue, DS> implements IBloc
     }
   }
 
-  private async updatePOI(height: number, blockHash: string, operationHash: Uint8Array): Promise<void> {
+  // First creation of POI
+  // TODO, add test here, even block been fetched, if no operation, we still treat it as empty block
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts:194]: no longer need

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 12]
```diff
@@ -1,55 +1,261 @@
...
 
-  async getLatestPoiBlockHash(): Promise<Uint8Array> {
-    if (!this.latestPoiBlockHash || !isMainThread) {
-      const poiBlockHash = await this.fetchPoiBlockHashFromDb();
-      if (poiBlockHash === null || poiBlockHash === undefined) {
-        this.latestPoiBlockHash = DEFAULT_PARENT_HASH;
+  async syncPoi(exitHeight?: number, logging?: boolean): Promise<void> {
+    if (this.isSyncing) return;
+    this.isSyncing = true;
+    while (!this.isShutdown) {
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:174]: Good point !

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 8 above)
Comment (stwiname) [reply packages/node-core/src/indexer/poi/poi.service.ts:80]: Maybe we can log that they should clean it up?

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (stwiname) [reply packages/node-core/src/indexer/poi/poi.service.ts:176]: I don't think we should have random delays like this. Its near impossible to follow code with things like that. e.g. if for some reason `ensureGenesisPoi` fails or a refactor means its not called we could be scratching our heads for hours trying to figure out why this isn't running. 

We're better off calling functions in the correct order

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 9 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:258]: the `setLatestSyncedPoi ` provide any option to force flush the cache, so we only await if we need to force flush.

When we create default poi blocks with a continuous range height, we can just use void here.

Comment (stwiname) [reply packages/node-core/src/indexer/poi/poi.service.ts:264]: Could we just set it at the end of the for loop?

Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:264]: it get parentHash from `this.latestSyncedPoi`, so it need to be set at every single block

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 3 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts:59]: We don't have any feature to reset syncedPoi, once we have this feature, we could log here let user to reset to previous valid poi height, also complete https://github.com/subquery/subql/blob/bfe6de09f6743db219f5e46f2022bf41711feb3a/packages/node-core/src/indexer/poi/poi.service.ts#L72

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/poi.service.ts]: Genesis Poi need to create one time only. 
Once in metadata stored `latestSyncedPoi` and in the `init` will set this object, the sync process will immediately start. 

 Before I think we could make sure sync process start only after when the non-null operation block been processed, rather than checking every 10 seconds, but we also could face an issue here: 
Assume we restart the app, in current poi, latestSyncedPoi is 1000, and we have poi 1002,1005 awaiting to be synced. 
Then after 1005 ,there is no blocks with operation. which means 1002 and 1005 will be never been synced.. 

In current implementation will solve this issue. 

I will leave a comment with this method

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 3 above)
Comment (ianhe8x) [reply packages/node-core/src/indexer/poi/poi.service.ts:59]: Can we treat this same as `_latestSyncedPoi` not exist?

File: packages/node-core/src/indexer/poi/PoiBlock.ts [hunk 13]
```diff
@@ -12,57 +12,84 @@ import {
...
   readonly id: number;
-  readonly chainBlockHash: Uint8Array;
-  readonly hash: Uint8Array;
-  readonly parentHash: Uint8Array;
-  readonly operationHashRoot: Uint8Array;
+  readonly chainBlockHash: Uint8Array | null;
+  readonly operationHashRoot: Uint8Array | null;
   readonly projectId: string;
+  readonly parentHash: Uint8Array | undefined;
+  readonly hash: Uint8Array | undefined;
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/PoiBlock.ts:48]: I found these names confusing, i thought they referred to the block hashes but they are actually the POI hashes.

File: packages/node-core/src/indexer/poi/PoiBlock.ts [hunk 14]
```diff
@@ -12,57 +12,84 @@ import {
...
 
   constructor(
     id: number,
-    chainBlockHash: Uint8Array,
-    hash: Uint8Array,
-    parentHash: Uint8Array,
-    operationHashRoot: Uint8Array,
-    projectId: string
+    chainBlockHash: Uint8Array | null,
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/PoiBlock.ts:52]: Could we change this to a string? Currently we need to ensure we can convert any block hash format to Uint8Array so when we add support for a new chain if they provide a different encoding we need to make changes here. 

It would help remove the big if/else block in create

File: packages/node-core/src/indexer/poi/poi.service.ts (hunk 6 above)
Comment (stwiname) [reply packages/node-core/src/indexer/poi/poi.service.ts:158]: So the code above this is used instead of this?

**Review by stwiname — COMMENTED** [review]
It would be good to have some documentation of how POI works, maybe just some comments in PoiBlock.ts 

The other concern I have is that we're computing POI for blocks that have no operations. Optimism/Arbitrum for example have 100M+ blocks. That is a large amount of data to store just for POI. This is due to wanting to query the POI at any block height. 

Is there something better we could do here? Like:
* Say the POI for any height is the same as the last height there was an operation
* Compute the POI at a specific height when requested as opposed to storing it and it never being queried. 
    * We could also have checkpoints every N blocks so we don't have to compute too much

File: packages/node-core/src/indexer/poi/PoiBlock.ts (hunk 13 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/PoiBlock.ts]: I added some comments for this

File: packages/node-core/src/indexer/poi/PoiBlock.ts (hunk 14 above)
Comment (jiqiang90) [reply packages/node-core/src/indexer/poi/PoiBlock.ts:58]: This will require convert any exiting chainBlockHash in db convert from `bytea` to `text`. It will increase complexity for migration, we can leave this now

File: packages/node-core/src/indexer/poi/poi.service.ts [hunk 15]
```diff
@@ -29,9 +30,12 @@ export class PoiService implements OnApplicationShutdown {
   private _poiRepo?: CachePoiModel;
   private _latestSyncedPoi?: ProofOfIndex;
   private isSyncing = false;
-  private _projectId?: string;
 
-  constructor(private storeCache: StoreCacheService, private eventEmitter: EventEmitter2) {}
+  constructor(
+    private storeCache: StoreCacheService,
+    private eventEmitter: EventEmitter2,
+    private project: ISubqueryProject
```
Comment (stwiname) [inline packages/node-core/src/indexer/poi/poi.service.ts:37]: You can use `@Inject('ISubqueryProject')` here instead of having to use a factory

**Review by stwiname — APPROVED** [review]
